llama-index-integrations/llms/llama-index-llms-bedrock-converse/llama_index/llms/bedrock_converse/base.py

```diff
@@ -0,0 +1,510 @@
+                            additional_kwargs={
+                                "tool_calls": tool_calls,
+                                "tool_call_id": tool_call_ids,
+                                "status": status,
+                            },
+                        ),
+                        raw=response,
+                    )
+
+        return gen()
+
+    @llm_completion_callback()
+    def stream_complete(
+        self, prompt: str, formatted: bool = False, **kwargs: Any
+    ) -> CompletionResponseGen:
+        stream_complete_fn = stream_chat_to_completion_decorator(self.stream_chat)
+        return stream_complete_fn(prompt, **kwargs)
+
+    @llm_chat_callback()
+    async def achat(
+        self, messages: Sequence[ChatMessage], **kwargs: Any
+    ) -> NotImplementedError:
+        raise NotImplementedError("Async chat is not supported for Bedrock Converse.")
+
+    @llm_completion_callback()
+    async def acomplete(
+        self, prompt: str, formatted: bool = False, **kwargs: Any
+    ) -> NotImplementedError:
+        raise NotImplementedError(
+            "Async completion is not supported for Bedrock Converse."
+        )
+
+    @llm_chat_callback()
+    async def astream_chat(
+        self, messages: Sequence[ChatMessage], **kwargs: Any
+    ) -> NotImplementedError:
+        raise NotImplementedError(
+            "Async stream chat is not supported for Bedrock Converse."
+        )
+
+    @llm_completion_callback()
+    async def astream_complete(
+        self, prompt: str, formatted: bool = False, **kwargs: Any
+    ) -> NotImplementedError:
+        raise NotImplementedError(
+            "Async stream completion is not supported for Bedrock Converse."
+        )
+
+    def chat_with_tools(
+        self,
+        tools: List["BaseTool"],
+        user_msg: Optional[Union[str, ChatMessage]] = None,
+        chat_history: Optional[List[ChatMessage]] = None,
+        verbose: bool = False,
+        allow_parallel_tool_calls: bool = False,
+        **kwargs: Any,
+    ) -> ChatResponse:
+        """Predict and call the tool."""
+        chat_history = chat_history or []
+
+        if isinstance(user_msg, str):
+            user_msg = ChatMessage(role=MessageRole.USER, content=user_msg)
+            chat_history.append(user_msg)
+
+        # convert Llama Index tools to AWS Bedrock Converse tools
+        tool_dicts = tools_to_converse_tools(tools)
+
+        response = self.chat(chat_history, tools=tool_dicts, **kwargs)
+
+        if not allow_parallel_tool_calls:
+            force_single_tool_call(response)
+
+        return response
+
+    async def achat_with_tools(
+        self,
+        tools: List["BaseTool"],
+        user_msg: Optional[Union[str, ChatMessage]] = None,
+        chat_history: Optional[List[ChatMessage]] = None,
+        verbose: bool = False,
+        allow_parallel_tool_calls: bool = False,
+        **kwargs: Any,
+    ) -> NotImplementedError:
+        raise NotImplementedError(
```

Comment:
I know it's a bit of a pain, but it would be a huge help to have async implemented here (using async boto3)

Reply:
Not sure how to do this, can you please suggest some resources or try to tackle this yourself?

Reply:
I don't have access to bedrock to test 😅 but I'm pretty sure its as simple as creating an async client with this no?
https://pypi.org/project/aioboto3/

Reply:
Should this async functionality be a separate PR? I'm asking because:
1. I would personally benefit from having synchronous integration of Bedrock Converse available now
2. The original Bedrock integration also doesn't have async implemented
3. There's ongoing work on figuring out how to use Bedrock async #10714; we can pull from the learnings there once that is concluded

Reply:
Fair enough. Note that it will be blocking for any feature that makes async llm calls though (tree summarize, sub question query engine, etc.)